### ISSUE TITLE
Add an "Optimization" recipe

### ISSUE THREAD

**Issue by markerikson** [reactjs/redux#1783]
Per discussion in #1751 , it's time we added a recipes page to cover performance and optimization.  @JAndritsch has volunteered to help write it, and I can definitely provide feedback and assistance.

Initial potential set of topics to cover per that discussion:

1. Connect components to state at the lowest level possible to prevent cascading renders from parent to children.
2. Normalize your data and pass collections of IDs instead of collections of large objects to components.
3. Use ImmutableJS.
4. Take advantage of "initialProps".
5. Re-define shouldComponentUpdate() when necessary.
6. Use React.addons.Perf to find and eliminate wasted renders.
7. Use Chrome's built-in profiling tools.
8. Memoization with Reselect
9. Per-instance component memoization
10. Performance *in general*. That subscriptions are O(n), that React Redux shallowly compares props, that memoization helps us avoid recalculations, etc.

Link-dumping a bunch of potential references, docs, and discussions:

- https://github.com/markerikson/react-redux-links/blob/master/react-performance.md
- http://somebody32.github.io/high-performance-redux/
- http://stackoverflow.com/questions/37264415/how-to-optimize-small-updates-to-props-of-nested-component-in-react-redux
- https://github.com/mweststrate/redux-todomvc/pull/1
- http://stackoverflow.com/questions/37285200/react-redux-performance-optimization-with-componentshouldupdate
- https://twitter.com/dan_abramov/status/720219615041859584

**Comment by markerikson** [comment]
Some relevant issues and docs:

- The "Performance" section in the FAQ: http://redux.js.org/docs/FAQ.html 
- https://github.com/reactjs/redux/issues/1603
- https://github.com/reactjs/redux/issues/1303

**Comment by JAndritsch** [comment]
Thanks for putting this together, Mark. I should have some free time this week to start combing through those links and gathering info. I'll ping you once I have a rough draft.

**Comment by somebody32** [comment]
Looks really good! I think that list is not in the final order of topics appearing in the guide?
And my 5 cents for now: I think it will be cool to add a topic about visualizing react component rerenders (with tools like [render-visualizer](https://github.com/spredfast/react-transform-render-visualizer)). I know that this is more general to react perf, but for me, it was really helpful to find wrong container/representation hierarchy on lists

**Comment by markerikson** [comment]
That was just the list of topics that @JAndritsch and @gaearon were proposing over in #1751 . Don't think any specific order was implied.  And yeah, render visualization and such would be good to cover too.

**Comment by JisuPark** [comment]
Why don't we list the helpful tools for optimization? 
In my case
- easy to handle shouldComponentUpdate with [Recompose](https://github.com/acdlite/recompose)
- babel plugin: [transform-react-constant-elements](https://babeljs.io/docs/plugins/transform-react-constant-elements/), [transform-react-inline-elements](http://babeljs.io/docs/plugins/transform-react-inline-elements/).
- eslint rule: [prefer-stateless-function](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prefer-stateless-function.md)

also useful.

**Comment by JAndritsch** [comment]
Update: I've got a good start on this documentation and am continuing to work on it when time is available. I'll hopefully have a draft for review soon.

**Comment by markerikson** [comment]
Cool, was going to ask if you'd had any time to look at it.  I still haven't managed to start in on the "Structuring Reducers" page myself.  Maybe this weekend, maybe?

Definitely ping me when you've got something to look at.

**Comment by HugoHeneault** [comment]
Any updates on this guys? :)

**Comment by cristian-sima** [comment]
@markerikson  Can you give a code example for each of the 10 points?

**Comment by markerikson** [comment]
@cristian-sima : I'm fairly busy at the moment, and when I do have time, my current priority is trying to work on the "Structuring Reducers" recipe described in #1784  .  I'd suggest reading through the links listed, particularly the set of articles included in the "React Performance" part of my React links list (which is the first link mentioned in this issue).

**Comment by cristian-sima** [comment]
@markerikson How many `connect` functions do you think I should use?

My app is going slow, and I am using just `5 connect`s

**Comment by HugoHeneault** [comment]
@cristian-sima have you checked memoization? 

http://redux.js.org/docs/recipes/ComputingDerivedData.html

**Comment by cristian-sima** [comment]
I have no idea what is that, but I will look :) 

Thanks

**Comment by naw** [comment]
My experience is [redux-devtools](https://github.com/gaearon/redux-devtools) can slow things down quite a bit.  If you are using redux-devtools, you might want to check how slow your app is _without_ it before you get too far down the optimization road. You should be able to use substantially more than 5 connects.

**Comment by cristian-sima** [comment]
A significant improvement in performance I've made today by setting a `shouldComponentUpdate`.

Essentially, `redux` makes all the components to update. So, telling which to update has helped me

**Comment by wub** [comment]
Excited for this recipe!

(If this <s>comment</s> novel should be its own issue or S/O question, let me know).

We're building an "activity feed" similar to Facebook's News Feed. We have hundreds of `activities` each with their own `connect`. Each activity has many sub-components (thumbnails that can be clicked on for modals, like/follow buttons etc.).

How should I `connect` each activity so we can optimise for `shouldComponentUpdate`? We've already normalised the data into `activities` (x liked 5 photos) and `activityObjects` (the 5 photos that x liked). Would one `connect` at the root `activity` suffice? Each `activityObject` can show up in multiple places in the feed. There are also `members` that can be followed (also normalised), and the indicator for that shows up in multiple places as well (for example when you click on a photo thumbnail `activityObject`).

Where do we hydrate the `activityObjects` for the best memoization? Do we denormalise at the root activity level and pass down all the props? Or `connect` each subcomponent _only_ to the parts of state that it cares about?

**Comment by markerikson** [comment]
Hi, @JAndritsch .  Wanted to check in and see if you are still planning to work on this.  You said you'd started a draft - is it posted anywhere?

**Comment by JAndritsch** [comment]
Hey @markerikson. 

Sorry for the delays here. Admittedly, I've had less time to work on this than originally thought. That being said, I still plan on putting in the effort when I can. I will send you the first draft when it's completed.

**Comment by markerikson** [comment]
Gotcha.  Yeah, I understand the "find time to write" issues :) 

I'd actually suggest pushing up your work in progress to a fork repo now. That way we can help contribute as you go. 

Thanks for helping with this!

**Comment by JAndritsch** [comment]
@markerikson Good idea. I created a separate branch and started the optimization recipe. Here's a direct link to it: https://github.com/JAndritsch/redux/blob/optimization-recipe/docs/recipes/Optimization.md.

**Comment by JAndritsch** [comment]
So I was just discussing Redux optimization today at work and that conversation reminded me about this thread. I'd still like to contribute, assuming the content is accurate. Would anyone here be interested in reviewing the draft above? If all looks well, I'll try to finish out the remaining sections and then submit a PR.

**Comment by markerikson** [comment]
Yeah, I'll try to take a look at it over the next day or two.

On a related note, I wrote a blog post a couple months ago that discussed some of the major aspects of performance: [Practical Redux, Part 6: Connected Lists, Forms, and Performance](http://blog.isquaredsoftware.com/2017/01/practical-redux-part-6-connected-lists-forms-and-performance/).  Might be helpful to cross-reference some of the ideas in there.
